### PR TITLE
(DOCSP-43059) Connect to MongoDB > Configure TLS

### PR DESCRIPTION
**Pull request by amalhotra-mdb**
# Pull Request Info

Create a child page on connecting to MongoDB that specifies how to configure TLS/SSL
Topics: TLS/SSL

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-NNNNN>
Staging - <https://docs-mongodbcom-staging.corp.mongodb.com/drivers/docsworker-xlarge/NNNNN/>

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?
- [ ] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/connect/tls.txt [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+- Uses TLS to connect to the MongoDB deployment
+- Verifies the deployment's certificate
+- Ensures that the certificate certifies the deployment
+
+To learn how to configure your MongoDB deployment for TLS, see the
+:manual:`TLS configuration guide </tutorial/configure-ssl/>` in the
+MongoDB Server manual. 
+
+.. important::
```
Comment (norareidy) [inline source/connect/tls.txt:36]: S: [nit] I feel like this admonition is more of a note than an important callout
```suggestion
.. note::
```

File: source/includes/connect/tls-ca-file.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include <mongoc/mongoc.h>
```
Comment (norareidy) [inline source/includes/connect/tls-ca-file.c:1]: S: you can move all these code examples in tls-ca-file.c, tls-connection-str.c, and tls-uri-options.c to one file and just separate each example with brackets (see the [replace documents PR](https://github.com/mongodb/docs-c/pull/52/files#diff-55e255805de91bb8bd44f27402c5e284f27b214062a8cd2e202ed9bd7751ee42) for an example of that)

File: source/includes/connect/tls-ca-file.c [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#include <bson/bson.h>
+
+int main (void) {
+
+  // start-connect
+  // Initialize the MongoDB C Driver
+  mongoc_init ();
```
Comment (norareidy) [inline source/includes/connect/tls-ca-file.c:8]: S: I don't think you need to include this in the code example; you can move mongoc_init() outside the //start-connect and //end-connect comments. If you want to make sure readers initialize the C driver before using it, you could add a note to the page that looks like the[ C++ driver instance note](https://www.mongodb.com/docs/languages/cpp/cpp-driver/upcoming/connect/client/#overview).

File: source/includes/connect/tls-ca-file.c [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  
+  // Create a new client, set TLS option to true, set path to client certificate file
```
Comment (norareidy) [inline source/includes/connect/tls-ca-file.c:10]: S: Since you're already describing the code in the rST, I'd remove this comment or move it outside the //start- and //end- comments

File: source/includes/connect/tls-ca-file.c [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  mongoc_uri_t *uri = mongoc_uri_new ("mongodb://localhost:27017");  
+  mongoc_uri_set_option_as_bool (uri, MONGOC_URI_TLS, true);
+  mongoc_uri_set_option_as_utf8 (uri, MONGOC_URI_TLSCERTIFICATEKEYFILE, "/path/to/client-certificate.pem");  
+  
+  // { Do database work here }
```
Comment (norareidy) [inline source/includes/connect/tls-ca-file.c:15]: S:
```suggestion
  // Do database work here
```

File: source/includes/connect/tls-ca-file.c [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  
+  // Cleanup
```
Comment (norareidy) [inline source/includes/connect/tls-ca-file.c:17]: S: I'd also remove this comment (we try to keep them to a minimum)
```suggestion
```

File: source/includes/connect/tls-ca-file.c (hunk 3 above)
Comment (norareidy) [reply source/includes/connect/tls-ca-file.c:8]: Applies to all the code files

File: source/includes/connect/tls-ca-file.c (hunk 4 above)
Comment (norareidy) [reply source/includes/connect/tls-ca-file.c:10]: Applies to all the code files

File: source/includes/connect/tls-ca-file.c (hunk 5 above)
Comment (norareidy) [reply source/includes/connect/tls-ca-file.c:15]: Applies to all

File: source/includes/connect/tls-ca-file.c (hunk 6 above)
Comment (norareidy) [reply source/includes/connect/tls-ca-file.c:17]: Applies to all

File: source/includes/connect/tls-connection-str.c [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+#include <bson/bson.h>
+
+int main (void) {
+
+  // start-connect-to-atlas
+  // Initialize the MongoDB C Driver
+  mongoc_init ();
+
+  // Create a new client and set tls option to true
+  mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname/port>/?replicaSet=myRepl&tls=true");
```
Comment (norareidy) [inline source/includes/connect/tls-connection-str.c:11]: Q: do you need to set the `replicaSet` option here? if so, I'd describe that in the text; otherwise you can remove that option

File: source/connect/tls.txt [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+   To learn more about the SRV connection format, see
+   :manual:`SRV Connection Format </reference/connection-string/#srv-connection-format>`
+   in the {+mdb-server+} documentation.
+
+.. _c-specify-ca-file:
+
+Specify a CA File
+------------------
+
+When MongoDB is started with TLS enabled, it will by default require the client to provide
```
Comment (norareidy) [inline source/connect/tls.txt:88]: S: since "MongoDB is started" is a little unclear, I'd reword to use active voice - maybe something like:
```suggestion
When you connect to a MongoDB deployment with TLS enabled, the deployment will by default require the client to provide
```

File: source/connect/tls.txt [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+a client certificate issued by a certificate authority specified by `--tlsCAFile`, or an authority
```
Comment (norareidy) [inline source/connect/tls.txt:89]: S: could you explain what specifying with `--tlsCAFile` means? or maybe just remove that part and make the wording a little more generic

```suggestion
a client certificate issued by a certificate authority, or an authority
```

File: source/connect/tls.txt [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+trusted by the native certificate store in use on the server.
+
+To provide the client certificate, set the `tlsCertificateKeyFile` in the URI to a PEM armored
```
Comment (norareidy) [inline source/connect/tls.txt:92]: S: since the code example is using the `mongoc_uri_set_option_as_utf8()` function to set the `MONGOC_URI_TLSCERTIFICATEKEYFILE` option, rather than including `tlsCertificateKeyFile` in the URI, I'd adjust the wording to reflect that:
```suggestion
To provide the client certificate, use the `mongoc_uri_set_option_as_utf8()` function to set the MONGOC_URI_TLSCERTIFICATEKEYFILE option to to a `.pem` file containing the root certificate chain:
```

Comment (norareidy) [reply source/connect/tls.txt:92]: Is there also a way to set this option directly in the URI?

File: source/connect/tls.txt [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+         :end-before: end-connect
+         :dedent:
+
+.. _c-certificate-revocation:
+
+Server Certificate Verification
+-------------------------------
+
+The MongoDB C Driver will automatically verify the validity of the server certificate, such as issued
+by configured Certificate Authority, hostname validation, and expiration.
```
Comment (norareidy) [inline source/connect/tls.txt:107]: S: (if I'm understanding this correctly)

```suggestion
The MongoDB C Driver will automatically verify the validity of a server certificate issued
by the configured Certificate Authority. The driver also performs hostname validation and revocation checking.
```

File: source/connect/tls.txt [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+It is not recommended to change these defaults as it exposes the client to *Man In The Middle* attacks (when
+``tlsAllowInvalidHostnames`` is set), invalid certificates (when ``tlsAllowInvalidCertificates`` is set), or
+potentially revoked certificates (when ``tlsDisableOCSPEndpointCheck`` or ``tlsDisableCertificateRevocationCheck``
+are set).
+
+Supported Libraries
+-------------------
+
+By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled
+by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. Valid values are:
```
Comment (norareidy) [inline source/connect/tls.txt:125]: S: clarify what these values are
```suggestion
by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:
```

File: source/connect/tls.txt [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+- ``AUTO`` the default behavior. Link to the system's native TLS library, or attempt to find OpenSSL.
```
Comment (norareidy) [inline source/connect/tls.txt:127]: S: Usually we use colons as [separators in lists](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items)

```suggestion
- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
```

File: source/connect/tls.txt [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+- ``DARWIN`` link to Secure Transport, the native TLS library on macOS.
```
Comment (norareidy) [inline source/connect/tls.txt:128]: S:
```suggestion
- ``DARWIN``: Links to Secure Transport, the native TLS library on macOS.
```

Comment (norareidy) [reply source/connect/tls.txt:128]: Applies to all list items

File: source/connect/tls.txt [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+OpenSSL
+~~~~~~~
+
+The MongoDB C Driver uses OpenSSL on Linux and Unix platforms (besides macOS). Industry best practices
+and some regulations require the use of TLS 1.1 or newer, which requires at least OpenSSL 1.0.1. Use the following command
+to check your OpenSSL version:
+
+.. code-block:: bash
+
+   $ openssl version
```
Comment (norareidy) [inline source/connect/tls.txt:143]: S: we try to avoid [using shell prompts](https://www.mongodb.com/docs/meta/style-guide/style/code-examples/#code-examples) ($) in the code:
```suggestion
   openssl version
```

File: source/connect/tls.txt [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+
+When compiled against OpenSSL, the driver will attempt to load the system default certificate store, as configured by the
+distribution. That can be overridden by setting the ``tlsCAFile`` URI option or with the fields ``ca_file`` and ``ca_dir``
+in the `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__.
+
+The Online Certificate Status Protocol (OCSP) (see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__) is fully supported
+when using OpenSSL 1.0.1+ with the following notes:
+
+- When a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
+  the server's certificate, the certificate is considered revoked (even if the certificate has a valid stapled OCSP response).
```
Comment (norareidy) [inline source/connect/tls.txt:160]: S: S: I'd introduce that external link and avoid having information in parentheses. Also, since there's only one note in the list, you can just combine the info into one paragraph:

```suggestion
The Online Certificate Status Protocol (OCSP) is fully supported
when using OpenSSL 1.0.1+. However, when a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
  the server's certificate, the certificate is considered revoked, even if the certificate has a valid stapled OCSP response.
  
 .. tip::
 
   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.
```

File: source/connect/tls.txt [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+
+When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
+and will issue an error if used.
+
+Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.
+
+The Online Certificate Status Protocol (OCSP) (see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__) is partially supported
+with the following notes:
+
+- The Must-Staple extension (see `RFC 7633 <https://tools.ietf.org/html/rfc7633>`__) is ignored.
```
Comment (norareidy) [inline source/connect/tls.txt:176]: S: same comment as above about introducing the external RFC links

File: source/connect/tls.txt [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+Native TLS Support on Windows (Secure Channel)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The MongoDB C Driver supports the Windows native TLS library (Secure Channel, or SChannel) and its native crypto library
+(Cryptography API: Next Generation, or CNG).
+
+When compiled against the Windows native libraries, the ``ca_dir`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported
+and will issue an error if used.
+
+Encrypted PEM files (e.g., setting ``tlsCertificateKeyPassword``) are also not supported and will result in error when
```
Comment (norareidy) [inline source/connect/tls.txt:189]: S: can you explain what the options you mention in this section are? I'm guessing URI options
```suggestion
Encrypted PEM files (e.g., setting the ``tlsCertificateKeyPassword`` URI option) are also not supported and will result in error when
```

File: source/connect/tls.txt [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+- When a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__, and the ``crl_file`` revokes the server's certificate, the OCSP response
+  takes precedence. For example, if the server presents a certificate with a valid stapled OCSP response, the certificate is considered
+  valid even if the ``crl_file`` marks it as revoked.
+
+.. _secure-transport:
+
+Native TLS Support on macOS / Darwin (Secure Transport)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The MongoDB C Driver supports the Darwin (OS X, macOS, iOS) native TLS library (Secure Transport), and its native crypto
```
Comment (norareidy) [inline source/connect/tls.txt:216]: S: I'm a little confused by the parentheses here; I think you can remove the "(OS X, macOS, iOS)" one, maybe explain briefly how "native TLS library relates" to "(Secure Transport)", and then just say "and Common Crypto, its native crypto library."

**Review by norareidy — CHANGES_REQUESTED** [review]
Nice work so far! l left some suggestions and questions

File: source/includes/connect/tls-connection-str.c (hunk 7 above)
Comment (amalhotra-mdb) [reply source/includes/connect/tls-connection-str.c:11]: nope not needed!

**Comment by amalhotra-mdb** [comment]
Thank you for your help on this @norareidy ! I was having some trouble parsing apart some of the denser content I cobbled together so I really appreciate your thoroughness!

File: source/connect/tls.txt [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+Specify a CA File
+------------------
+
+When you connect to a MongoDB deployment with TLS enabled, the deployment will by default require the client to provide
+a client certificate issued by a certificate authority, or an authority
+trusted by the native certificate store in use on the server.
+
+You can provide the client certificate in the following ways:
+
+- Setting the ``tlscertificatekeyfile`` parameter in your connection string to a PEM armored certificate file
```
Comment (norareidy) [inline source/connect/tls.txt:94]: S: I'd use the same description of a pem file in both these bullet points (and use the same placeholder value in the code tabs)
```suggestion
- Setting the ``tlscertificatekeyfile`` parameter in your connection string to a ``.pem`` file containing the root certificate chain
```

File: source/connect/tls.txt [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+.. _c-certificate-revocation:
+
+Server Certificate Verification
+-------------------------------
+
+The MongoDB C Driver will automatically verify the validity of a server certificate issued
+by the configured Certificate Authority. The driver also performs hostname validation and revocation checking.
+
+To overwrite this behavior, it is possible to disable hostname validation, OCSP endpoint revocation checking,
+revocation checking entirely, and allow invalid certificates.
```
Comment (norareidy) [inline source/connect/tls.txt:127]: S:
```suggestion
all revocation checking, and allow invalid certificates.
```

File: source/connect/tls.txt [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+
+This behavior is controlled using the ``tlsAllowInvalidHostnames``, ``tlsDisableOCSPEndpointCheck``,
+``tlsDisableCertificateRevocationCheck``, and ``tlsAllowInvalidCertificates`` options. By default,
+all are set to ``false``.
+
+It is not recommended to change these defaults as it exposes the client to *Man In The Middle* attacks (when
```
Comment (norareidy) [inline source/connect/tls.txt:133]: S: maybe use an inline link to a wikipedia page describing man in the middle attacks instead of italics, since readers might not know what that is

Comment (norareidy) [inline source/connect/tls.txt:133]: S: I think this paragraph would be more readable as a list; you could also avoid parentheses that way. Something like: "It is not recommended to change these defaults, since you might expose your client to the following security risks:
- Man in the middle attacks, when ``tlsAllowInvalidHostnames`` is set
...

File: source/connect/tls.txt [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+``tlsAllowInvalidHostnames`` is set), invalid certificates (when ``tlsAllowInvalidCertificates`` is set), or
+potentially revoked certificates (when ``tlsDisableOCSPEndpointCheck`` or ``tlsDisableCertificateRevocationCheck``
+are set).
+
+Supported Libraries
+-------------------
+
+By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled
+by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:
+
```
Comment (norareidy) [inline source/connect/tls.txt:143]: S: [nit] I'd list these values in the same order as the subsections appear

File: source/connect/tls.txt [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
+- ``DARWIN``: Links to Secure Transport, the native TLS library on macOS.
+- ``WINDOWS``: Links to Secure Channel, the native TLS library on Windows.
+- ``OPENSSL``: Links to OpenSSL (libssl). An optional install path may be specified with ``OPENSSL_ROOT``.
+- ``LIBRESSL``: Links to LibreSSL's libtls. (LibreSSL's compatible libssl may be linked to by setting ``OPENSSL``).
```
Comment (norareidy) [inline source/connect/tls.txt:148]: S: I'd reword to active voice and remove the parentheses
```suggestion
- ``LIBRESSL``: Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.
```

File: source/connect/tls.txt [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+LibreSSL / libtls
+~~~~~~~~~~~~~~~~~
+
+The MongoDB C Driver supports LibreSSL through the use of OpenSSL compatibility checks when configured to compile against
+``openssl``. It also supports the new ``libtls`` library when configured to build against ``libressl``.
+
+When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
+and will issue an error if used.
+
+Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.
```
Comment (norareidy) [inline source/connect/tls.txt:190]: Q: setting these options has no effect when compliling against the Windows native libraries, or always? If it's the former, I'd combine this sentence with the paragraph before

File: source/connect/tls.txt [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+Native TLS Support on Windows (Secure Channel)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The MongoDB C Driver supports the Windows native TLS library (Secure Channel, or SChannel) and its native crypto library
+(Cryptography API: Next Generation, or CNG).
+
+When compiled against the Windows native libraries, the ``ca_dir`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported
+and will issue an error if used.
+
+Encrypted PEM files (e.g., setting the ``tlsCertificateKeyPassword`` URI option) are also not supported and will result in error when
```
Comment (norareidy) [inline source/connect/tls.txt:212]: S: avoiding [parentheses](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#parentheses):
```suggestion
Encrypted PEM files, set by using the ``tlsCertificateKeyPassword`` URI option, are also not supported and will result in error when
```

File: source/connect/tls.txt [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+attempting to load them.
+
+When ``tlsCAFile`` is set, the driver will only allow server certificates issued by the authority (or authorities) provided.
```
Comment (norareidy) [inline source/connect/tls.txt:215]: S: parentheses
```suggestion
When ``tlsCAFile`` is set, the driver will only allow server certificates issued by one or more authorities provided.
```

**Review by norareidy — CHANGES_REQUESTED** [review]
Thanks so much for picking up this complicated page! Almost there, I just have a few more suggestions/questions

**Review by norareidy — APPROVED** [review]
LGTM!

File: source/connect/tls.txt [hunk 28]
```diff
@@ -0,0 +1,271 @@
+.. c-tls:
+
+========================================
+Configure Transport Layer Security (TLS)
```
Comment (kevinAlbs) [inline source/connect/tls.txt:4]: Confirming: is https://www.mongodb.com/docs/languages/c/c-driver/current/libmongoc/guides/configuring_tls/ planned to be removed?

File: source/connect/tls.txt [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+
+Supported Libraries
+-------------------
+
+By default, libmongoc will attempt to find a supported TLS library and enable TLS support. This is controlled
+by the cmake flag ``ENABLE_SSL``, which is set to ``AUTO`` by default. This flag accepts the following values:
+
+- ``AUTO``: Links to the system's native TLS library, or attempts to find OpenSSL. This is the default value.
+- ``OPENSSL``: Links to OpenSSL (libssl). An optional install path may be specified with ``OPENSSL_ROOT``.
+- ``LIBRESSL``: Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.
```
Comment (kevinAlbs) [inline source/connect/tls.txt:148]:
```suggestion
- ``LIBRESSL`` (Deprecated):  Links to LibreSSL's libtls. You can link to LibreSSL's compatible libssl by setting ``OPENSSL``.
```

Suggest noting LIBRESSL is deprecated. [C driver 1.28.0](https://github.com/mongodb/mongo-c-driver/releases/tag/1.28.0) deprecates use of LibreSSL:

> `ENABLE_SSL=LIBRESSL` is deprecated along with `mongoc_stream_tls_libressl_new`. Support for LibreSSL may be dropped in a future major release.

File: source/connect/tls.txt [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+The Online Certificate Status Protocol (OCSP) is fully supported
+when using OpenSSL 1.0.1+. However, when a ``crl_file`` is set with `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ and the ``crl_file`` revokes
+the server's certificate, the certificate is considered revoked, even if the certificate has a valid stapled OCSP response.
+  
+.. tip::
+ 
+   For more information about OCSP, see `RFC 6960 <https://tools.ietf.org/html/rfc6960>`__.
+
+LibreSSL / libtls
+~~~~~~~~~~~~~~~~~
```
Comment (kevinAlbs) [inline source/connect/tls.txt:184]:
```suggestion
LibreSSL / libtls (Deprecated)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: source/connect/tls.txt [hunk 31]
```diff
@@ -0,0 +1,271 @@
...
+
+The MongoDB C Driver supports LibreSSL through the use of OpenSSL compatibility checks when configured to compile against
+``openssl``. It also supports the new ``libtls`` library when configured to build against ``libressl``.
+
+When compiled against the Windows native libraries, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
```
Comment (kevinAlbs) [inline source/connect/tls.txt:189]:
```suggestion
When compiled with LibreSSL, the ``crl_file`` option of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ is not supported,
```

File: source/connect/tls.txt [hunk 32]
```diff
@@ -0,0 +1,272 @@
...
+The MongoDB C Driver supports both the Darwin native TLS library and Common Crypto, its native crypto
+library.
+
+When compiled against Secure Transport, the ``ca_dir`` and ``crl_file`` options of a `mongoc_ssl_opt_t <{+api-libmongoc+}/mongoc_ssl_opt_t.html>`__ are not supported.
+An error is issued if either are used.
+
+When ``tlsCAFile`` is set, the driver will only allow server certificates issued by the authority (or authorities) provided.
+When no ``tlsCAFile`` is set, the driver will use the Certificate Authorities in the unlocked keychains.
+
+Setting ``tlsDisableOCSPEndpointCheck`` and ``tlsDisableCertificateRevocationCheck`` has no effect.
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply source/connect/tls.txt:251]: When compiling against secure transport ([checked here](https://github.com/mongodb/mongo-c-driver/blob/56cbe883be2fe84812e35ea8fae181dc7e0be8ce/src/libmongoc/src/mongoc/mongoc-stream-tls-secure-transport.c#L695-L698)).

File: source/includes/connect/tls.c [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+#include <bson/bson.h>
+
+int main (void) {
+
+  mongoc_init ();
+
+  {
+    // Set TLS using connection string
+    // start-connect-str
+    mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname/port>/?tls=true");
```
Comment (kevinAlbs) [inline source/includes/connect/tls.c:11]:
```suggestion
    mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname>/?tls=true");
```

Quoting [Initial DNS Seedlist Discovery](https://github.com/mongodb/specifications/blob/c5dcae52a77d1ac90ca8c0cea7f371b73982062e/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.md):
> It is an error to specify a port in a `mongodb+srv` connection string.

File: source/includes/connect/tls.c [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+    
+    mongoc_client_destroy (client);
+    mongoc_uri_destroy (uri);
+    // end-connect-uri-opt
+  }
+
+  {
+    // Set TLS and provide CA file using connection string
+    // start-connect-ca-file-str
+    mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname/port>/?tls=true&tlscertificatekeyfile=/path/to/certs/client-certificate.pem");
```
Comment (kevinAlbs) [inline source/includes/connect/tls.c:37]:
```suggestion
    mongoc_client_t *client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname>/?tls=true&tlscertificatekeyfile=/path/to/certs/client-certificate.pem");
```

File: source/connect/tls.txt (hunk 28 above)
Comment (amalhotra-mdb) [reply source/connect/tls.txt:4]: yup – the docs are being standardized to the layout that other drivers docs follow, like the [pymongo docs](https://www.mongodb.com/docs/languages/kotlin/kotlin-sync-driver/current/indexes/), so this page will replace the previous 'configuring TLS' page

**Review by kevinAlbs — APPROVED** [review]
